sdk/trace/span.go

```diff
@@ -324,14 +326,42 @@ func truncateAttr(limit int, attr attribute.KeyValue) attribute.KeyValue {
 		v := trucated.Value.AsStringSlice()
 		for i := range v {
 			if len(v[i]) > limit {
-				v[i] = v[i][:limit]
+				v[i] = safeTruncate(v[i], limit)
 			}
 		}
 		return trucated
 	}
 	return attr
 }
 
+// safeTruncate truncates the string and guarantees valid UTF-8 is returned.
+func safeTruncate(input string, limit int) string {
+	if trunc, ok := safeTruncateValidUTF8(input, limit); ok {
+		return trunc
+	}
+	trunc, _ := safeTruncateValidUTF8(strings.ToValidUTF8(input, ""), limit)
+	return trunc
+}
+
+// safeTruncateValidUTF8 returns a copy of the input string safely truncated to
+// limit. The truncation is ensured to occur at the bounds of complete UTF-8
+// characters. If invalid encoding of UTF-8 is encountered, input is returned
+// with false, otherwise, the truncated input will be returned with true.
+func safeTruncateValidUTF8(input string, limit int) (string, bool) {
+	for cnt := 0; cnt <= limit; {
+		r, size := utf8.DecodeRuneInString(input[cnt:])
+		if r == utf8.RuneError {
+			return input, false
+		}
+
+		if cnt+size > limit {
```

Comment:
One thing I noted [from the spec](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute-limits) is that the limit is defined by characters not bits. I think we may need to keep a separate count of runes here, right?

Reply:
Don't think this is blocking, as we currently do not do this either. :shrug:

Reply:
Since this issue has popped up in a lot of places, I'd like to see a quick fix put in place while this gets debated. I'd prefer to see "bytes" be the definition, but I agree it's ill-specified.

Reply:
I would add that UTF-8 is documented as "self-synchronizing", which means we can implement an O(1) test for a valid UTF-8 boundary, if we want to, which may address part of our problem in a better way, but only part of the problem.

Given the discussion in https://github.com/open-telemetry/opentelemetry-proto/issues/426, I'm inclined to suggest that we: (1) not go after the O(1) test, (2) release this fix quickly, (3) debate for the next release what other UTF-8 sanitization should be done.

From anecdotal evidence, it's only the Golang protobuf library that's acting this way. We have seen C++ OTel clients sending invalid UTF-8 that we suspect also results from truncation, but in that case we see it when deserializing the data on our Go servers.

Reply:
@jmacd can you create an issue to track this in the next release?

Reply:
https://github.com/open-telemetry/opentelemetry-go/issues/3160